### DISCUSSION TITLE
Future Storage Layout

### DISCUSSION THREAD

**Discussion by lordfolken** [Openvario/meta-openvario#239] [Ideas]
This is currently a placeholder for the storage layout discussion.

**Comment by lordfolken** [comment]
/home partition. 
 * indepdendent of system partition /
 * should survive an upgrade even with a new image
 * should enlarge itsefl at boot to fill a certain percentage of the storage device.

**Reply by MaxKellermann** [reply]
This should not be mounted on `/home`; when an OpenVario user puts the SD card into his Windows computer, he shouldn't see all the gory details of a Linux home directory, no `.xcsoar` folder. Ideally that partition contains just an `XCSoarData` folder (instead of `.xcsoar`), like on most other XCSoar machines (Android, Kobo, and previously WinCE).

As the file system, we have several choices:
- `ext4` - best quality and robustness, but can only be accessed by Linux
- `fat` - compatible with everything, but ancient and bad, not acceptable
- `ntfs` - no, please not ...
- `udf` - compatible with everything (https://en.wikipedia.org/wiki/Universal_Disk_Format)
- `exfat` - compatible with everything (https://en.wikipedia.org/wiki/ExFAT)

If Windows compatibility is important, only `udf` and `exfat` are usable.

**Reply by lordfolken** [reply]
No other vario brand allows you to mount the internal storage on a regular windows machine. 
Underminding the storage design for that one potential usecase doesn't seem prudent.
Please just don't give users this option. it only leads to problems. The only way in/out should be either network or removable storage (excluding the os sdcard.)

**Reply by MaxKellermann** [reply]
I'd be very happy *not* to give Windows users access, because that complicates things - but the discussion on https://github.com/Openvario/meta-openvario/issues/156 about it did not lead to a final decision about it. If it is decided that Windows compatibility is not desired, we pick, of course, ext4.

But in any case, the contents of that data partition need to be "stable", as in: if you replace the root+boot partition with a new (or older?) version, there must not be any problems. No "oops there's a .bashrc which is not compatible with the new image version" and no "this ~root home folder does not contain files required by the new version".
This needs to "just work".
Therefore, designing this data partition properly *now* and documenting it *now*, before we start having it, is important. All of its contents must be explicit design, not "let's just mount this on /home and stuff starts landing there".

**Reply by lordfolken** [reply]
Actually if we do this clever, we could mount an usbstick on the userdata directory, and run directly from the stick.

Although i might have some reservations when we append log files and mount the stick synchronosly.

**Reply by kedder** [reply]
USB sticks are flimsy, I'd not trust a usb-socket in flight :)

**Reply by mihu-ov** [reply]
The "runnning from USB drive" discussion is a regular one, like watching Dinner for Two on new years eve ;-)
I think we already have a consensus on this. If anyone of the experts wants to do it he probably knows the limitations and risks. If we support that for the standard user there will be data loss, frustration and an exploding XCSoar forum.

**Reply by mihu-ov** [reply]
Once installed in a glider physical access to the µSD card is a real pain. That´s why I think other than bench testing there is no real use case for accessing the data partition from Windows. Easy options for data transfer from USB drive, smartphone, etc, are more important.

**Comment by MaxKellermann** [comment]
Sub-thread: what data really needs to be on the `data` partition?
- `XCSoarData`
- but what else?

Any other configuration or user data?

Machine-specific configuration is on the `boot` partition, not `data`, where it will be accessible by u-boot (`config.uEnv`, e.g. display rotation). This will be overwritten with a new image; but we could have a shadow copy on the new `data` partition, to be moved over on the first boot, so those settings are preserved?

**Reply by lordfolken** [reply]
There is some machine specific data, such as 
 * network settings (wifi/usb/rj45)
 * brightness / rotation / language / calibration
Whatever is settable by ovmenu or its replacement.

Ideally all that should persist over a new image.
but should be resettable / erasable to default. 

Maybe put it in a spearate folder ```OVData/``` along with ```XCSoarData/``` ?

**Reply by kedder** [reply]
There's a bunch of user-configurable data that ideally would've been preserved: configuration for the sensord/variod, touchscreen calibration data, wifi hotspots and their passwords. Obviously, 3rd party software usually expect to store their setting / state under $HOME.

**Reply by MaxKellermann** [reply]
> Obviously, 3rd party software usually expect to store their setting / state under $HOME.

And in `/etc`. And in `/var`.
But if you put those on the data partition, you'll run into lots of problems.

For our own software and other software we ship in the base image (variod, sensord etc.) we can easily put those in `OVData` as folken suggested, but suporting arbitrary other software will be difficult.

**Reply by kedder** [reply]
I agree, preserving `/etc` and `/var` _in general_ would not work well. Most of software is configurable in this regard  though, so if we designate a well known "persistent" directory, we can configure things that we care about to put persistent data there.

**Reply by lordfolken** [reply]
Well a lot of this speaks for running xcsoar under a  non privileged  user. That way its very clear where writing can happen and where it can't. 
Some enduser network tools can be configured in userspace olny.

**Reply by lordfolken** [reply]
For example wpa_supplicant is user capable.

**Reply by lordfolken** [reply]
i think we should only persist settings that we are directly responsible for. so stuff set via ovmenu mainly.

**Reply by MaxKellermann** [reply]
> Well a lot of this speaks for running xcsoar under a non privileged user.

Agree, see https://github.com/Openvario/meta-openvario/pull/190#discussion_r786909007

There are many reasons why XCSoar shouldn't be running as root; this is not about security (see @kedder's disagreement). For me, this is about correctness. Give XCSoar only the privileges it really needs, but not more. This is a bit more work, because we need to design the system properly, but in the end, we have a better design.

For example, don't `mount /dev/mmcblk0p1 /boot`; use the systemd automounter (or systemd dependencies) instead. As a non-root user, you can't just `mount`, and this restriction leads the path to the better solution.

(This is different from the Kobo, where XCSoar runs as root, too: the Kobo has no userspace, no systemd, no init at all; it boots right into KoboMenu/XCSoar. On the Kobo, we can't just create new users, because we live inside the existing Kobo firmware. But the OpenVario firmware can be designed properly.)

**Reply by kedder** [reply]
I hate to go into these philosophical discussions, because they are never conclusive, but.

> this is not about security [...] For me, this is about correctness.

Let's not forget _why_ it is considered a good practice to run things as unprivileged user. It is exclusively because this approach provides some additional security protections. All these different users and permissions exists _only_ to prevent someone else (logged on the system) from messing with your sensitive data. Openvario is effectively single-user device and this logic is not quite applicable: there is noone else logged on the system and there is no sensitive data to protect from them.

When you build and run xcsoar on your laptop, you are not running it as dedicated "xcsoar" user because it would be more correct. You are running it as the same user as yourself, because you also want to read the data it produces and you want to let xcsoar read your files. Openvario is a single user device, and this single user can be unpriviledged "john" or can be priviledged "root" - there is no much difference in practice, because there's no data to protect from it.

So, if we going to go this route, let explicitly why it should be "more correct", because the usual reasoning is simply not applicable for this particular case. Otherwise we're just going to burn energy on something that doesn't really makes sense.

**Reply by mihu-ov** [reply]
WiFi passwords are probably the only sensitive data I could think of.

**Reply by MaxKellermann** [reply]
> > this is not about security [...] For me, this is about correctness.
> 
> Let's not forget _why_ it is considered a good practice to run things as unprivileged user. It is exclusively because this approach provides some additional security protections. All these different users and permissions exists _only_ to prevent someone else (logged on the system) from messing with your sensitive data.

This discussion cannot possibly be conclusive or even lead *anywhere* if you ignore my statement explaining what this is **not** about, and then say this is **only about exactly that**. Even after citing the very statement you ignored, the one you contradicted with.

**Reply by lordfolken** [reply]
The main reason is and has always been: to protect the os from the application. 
For example: 
* The diskspace for the root user is usally reserved 10%. So even if the app fills the disk, the os can still boot and proceed. 
* The priortiy level can be overruled by an operating system process. So even if the app is consuming all cpu, the os is still responsive.
* No critical files can be damaged, by a fat finger or a bug in the app trying to write somewhere. (we have had such cases in xcsoar.) 

The whole idea is that as long the box is there, the situation can be remidiated. If its dead, another level of interference is necessary.

**Reply by August2111** [reply]
> Ideally that partition contains just an `XCSoarData` folder (instead of `.xcsoar`), like on most other XCSoar machines (Android, Kobo, and previously WinCE).

This I prefere: A uniform name for all distris also in Linux: XCSoarData...
And what I really want: If a user put his sd card in a Windows PC he can can easely cleanup this folder on this without any moving to another device! Maybe I'm a little bit old school for all developer here - but this it's real live on the airfields! Of course it is easier to handle a linux format like ext4 - but I also want to compare with other varios and so on.
It is also possible to split it in two folders, one for cache and logging - and another one for the data, saved flights and settings - but this makes additional work too.
As Max already explained: Certain settings cannot be easily transferred to an (external) folder - such as the display, touchscreen, wifi settings and more. And of course it would be very nice if these were retained after an update / new image...

**Reply by MaxKellermann** [reply]
> As Max already explained: Certain settings cannot be easily transferred to an (external) folder - such as the display, touchscreen, wifi settings and more. And of course it would be very nice if these were retained after an update / new image...

No, I did not say that. We can put anything we want on the data partition, but it must be an explicit decision with an explicit design. We can do that, and of course we should do that. No setting any (regular) user ever changes must be lost. The root partition should ideally be mounted read-only (for regular users).

Power users can, of course, have a writable root partition and hack everything - but a new firmware image will lose all of that. And yes, they can do opkg upgrades if they want and have a opkg server, instead of applying a whole new firmware image (but I do not want to support that).
